ledger/complete/wal/checkpoint_v6_reader.go

```diff
@@ -325,22 +325,63 @@ func readSubTriesConcurrently(dir string, fileName string, subtrieChecksums []ui
 	return nodesGroups, nil
 }
 
+func readCheckpointSubTrie(dir string, fileName string, index int, checksum uint32, logger *zerolog.Logger) (
+	[]*node.Node,
+	error,
+) {
+	var nodes []*node.Node
+	err := processCheckpointSubTrie(dir, fileName, index, checksum, logger,
+		func(reader *Crc32Reader, scratch []byte, nodesCount uint64) error {
+			nodes = make([]*node.Node, nodesCount+1) //+1 for 0 index meaning nil
+			logging := logProgress(fmt.Sprintf("reading %v-th sub trie roots", index), int(nodesCount), logger)
+			for i := uint64(1); i <= nodesCount; i++ {
+				node, err := flattener.ReadNode(reader, scratch, func(nodeIndex uint64) (*node.Node, error) {
+					if nodeIndex >= i {
+						return nil, fmt.Errorf("sequence of serialized nodes does not satisfy Descendents-First-Relationship")
+					}
+					return nodes[nodeIndex], nil
+				})
+				if err != nil {
+					return fmt.Errorf("cannot read node %d: %w", i, err)
+				}
+				nodes[i] = node
+				logging(i)
+			}
+			return nil
+		})
+
+	if err != nil {
+		return nil, err
+	}
+
+	// since nodes[0] is always `nil`, returning a slice without nodes[0] could simplify the
+	// implementation of getNodeByIndex
+	// return nodes[1:], nil
+	return nodes[1:], nil
+}
+
 // subtrie file contains:
 // 1. checkpoint version
 // 2. nodes
 // 3. node count
 // 4. checksum
-func readCheckpointSubTrie(dir string, fileName string, index int, checksum uint32, logger *zerolog.Logger) (
-	subtrieRootNodes []*node.Node,
+func processCheckpointSubTrie(
```

Comment:
extracting this function to be used by both existing reader and leaf nodes reader